native/cocos/network/DownloaderImpl-apple.mm

```diff
@@ -180,23 +189,57 @@ - (void)dealloc {
 @end
 
 @implementation DownloaderAppleImpl
-
 - (id)init:(const cc::network::DownloaderApple *)o hints:(const cc::network::DownloaderHints &)hints {
     DLLOG("Construct DownloaderAppleImpl %p", self);
     // save outer task ref
     _outer = o;
     _hints = hints;
-
+    
     // create task dictionary
     self.taskDict = [NSMutableDictionary dictionary];
 
     // create download session
-    NSURLSessionConfiguration *defaultConfig = [NSURLSessionConfiguration defaultSessionConfiguration];
-    self.downloadSession = [NSURLSession sessionWithConfiguration:defaultConfig delegate:self delegateQueue:[NSOperationQueue mainQueue]];
+//     NSURLSessionConfiguration *defaultConfig = [NSURLSessionConfiguration defaultSessionConfiguration];
+//     self.downloadSession = [NSURLSession sessionWithConfiguration:defaultConfig delegate:self delegateQueue:[NSOperationQueue mainQueue]];
     //    self.downloadSession.sessionDescription = kCurrentSession;
+
+    // use backgroundSession to support background download
+    self.downloadSession = [self backgroundURLSession];
+
+    self.hasUnfinishedTask = YES;
+    // cancel and save last running tasks
+    [self.downloadSession getTasksWithCompletionHandler:^(NSArray *dataTasks, NSArray *uploadTasks, NSArray *downloadTasks) {
+        for (NSURLSessionDownloadTask *downloadTask in downloadTasks) {
+            // cancelTask with resume Data
+            if ((long)downloadTask.state == 0) {
+                [(NSURLSessionDownloadTask *)downloadTask cancelByProducingResumeData:^(NSData *resumeData) {
+                    if(downloadTask.originalRequest.URL.absoluteString) {
+                        NSString *tempFilePathWithHash = [NSString stringWithFormat:@"%s%lu%s", cc::FileUtils::getInstance()->getWritablePath().c_str(), (unsigned long)[downloadTask.originalRequest.URL.absoluteString hash], _hints.tempFileNameSuffix.c_str()];
+                        [resumeData writeToFile:tempFilePathWithHash atomically:YES];
+                    } else if (downloadTask.currentRequest.URL.absoluteString) {
+                            NSString *tempFilePathWithHash = [NSString stringWithFormat:@"%s%lu%s", cc::FileUtils::getInstance()->getWritablePath().c_str(), (unsigned long)[downloadTask.currentRequest.URL.absoluteString hash], _hints.tempFileNameSuffix.c_str()];
+                        [resumeData writeToFile:tempFilePathWithHash atomically:YES];
+                    }
+                }];
+            }
+        }
+    }];
+    
     return self;
 }
 
+#pragma mark - backgroundURLSession
+static int tag = 0;
```

Comment:
Can skip using static variable here? It is better to define as class member variable.

Reply:
> Can skip using static variable here? It is better to define as class member variable.

No, it shouldn't be. Because backgroundSession's Identifier should be unique in global.

Reply:
How about static class member variable?

Reply:
> How about static class member variable?
Have been moved into the member function body.